plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```
@@ -3756,6 +3760,42 @@ public Optional<ConnectorOutputMetadata> finishRefreshMaterializedView(
+            TrinoFileSystem fileSystem = fileSystemFactory.create(session.getIdentity(), icebergTable.io().properties());
+            List<Location> pathsToDelete = new ArrayList<>();
+            // deleteFunction is not accessed from multiple threads unless org.apache.iceberg.ExpireSnapshots.executeDeleteWith is used
+            Consumer<String> deleteFunction = path -> {
+                pathsToDelete.add(Location.of(path));
+                if (pathsToDelete.size() == DELETE_BATCH_SIZE) {
+                    try {
+                        fileSystem.deleteFiles(pathsToDelete);
+                        pathsToDelete.clear();
+                    }
+                    catch (IOException e) {
+                        throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Failed to delete files during materialized view refresh", e);
+                    }
+                }
+            };
+
+            IcebergTableHandle icebergTableHandle = (IcebergTableHandle) tableHandle;
+            Table updatedIcebergTable = catalog.loadTable(session, icebergTableHandle.getSchemaTableName());
+            try {
+                updatedIcebergTable
+                        .expireSnapshots()
+                        .expireSnapshotId(snapshotIdBeforeRefreshCommit)
+                        // Explicitly set `.expireOlderThan` to current time otherwise snapshots from last 5 days are still retained as per https://github.com/apache/iceberg/blob/6e2eaf02f380ed8e3be7f58273d680b766897b18/core/src/main/java/org/apache/iceberg/RemoveSnapshots.java#L320
+                        .expireOlderThan(System.currentTimeMillis())
+                        .deleteWith(deleteFunction)
+                        .commit();
+
+                fileSystem.deleteFiles(pathsToDelete);
+            }
+            catch (IOException e) {
+                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Failed to delete files during snapshot expiration", e);
```

The transaction is already committed at L3767. I'm wondering if we really need to throw an exception here.

---

Good thought, we could change it to log.info or if we throw then, make the message little clear that the Trino couldn't collect the left over files from the refresh materialized view action.